rfcs/2020-09-21-0000-kubernetes-metrics-overview.md

```diff
@@ -0,0 +1,283 @@
+# RFC 0000 - 2020-09-21 - Overview of metrics in Kubernetes
+
+This RFC's goal is to create a systematic overview of metrics in the
+Kubernetes ecosystem.
+
+## Table of contents
+
+- [RFC 3684 - 2020-09-21 - Overview of metrics in Kubernetes](#rfc-3684---2020-09-21---overview-of-metrics-in-kubernetes)
+  - [Table of contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Overview](#overview)
+  - [Metrics exposed by the Kubernetes system components](#metrics-exposed-by-the-kubernetes-system-components)
+    - [Properties](#properties)
+    - [Per-component insights](#per-component-insights)
+  - [TODO](#todo)
+  - [Prior art](#prior-art)
+
+## Motivation
+
+Metrics is Kubernetes are a large sector, with many concerns. We need a way to
+better understand it, to, sort of, have a list of entry points to explore in
+more detail.
+
+This overview will help us reason about the metrics  in Kubernetes and make
+educated design and planning decisions.
+
+## Overview
+
+- Application metrics exposed via [OpenMetrics](https://openmetrics.io/) or
+  [Prometheus exposition format](https://prometheus.io/docs/instrumenting/exposition_formats/)
+  for scraping, aka pull-based gathering.
+
+- Application metrics shipped to a server of some kind, typically via `statsd`
+  protocol, aka push-based gathering.
+
+  In the modern metrics aggregation stacks, these are usually converted to
+  pull-based gathering via
+  [pushgateway](https://github.com/prometheus/pushgateway) or
+  [statsd_exporter](https://github.com/prometheus/statsd_exporter) - but
+  it's even more common to not have any push-based metrics at all.
+
+- Metrics exposed by the Kubernetes system components via Prometheus exposition
+  format.
+
+  Among others, this category includes:
+
+  - `kubelet`
+  - `kube-apiserver`
+  - `kube-scheduler`
+  - `kube-controller-manager`
+  - `etcd`
+  - `docker`
+
+  More on this [later](#metrics-exposed-by-the-kubernetes-system-components),
+  but the key part here is those components are not all deployed together and
+  require different access paths if we want to reach their `/metrics` endpoints.
+
+- Metrics available via the [Metrics API] - the `kube-apiserver` API that offers
+  system metrics for the purposes of autoscaling to various Kubernetes
+  components (typically autoscalers -  i.e. Horizontal Pod Autoscaler,
+  Vertical Pod Autoscaler, etc).
+
+  This API essentially contains very limited data on resource usage of the
+  `Pod`s and `Node`s.
+
+  Reviewed in more detail at the Metrics API RFC.
+
+- Metrics gathered from the kernel, at the host level (CPU/RAM/IO usage).
```

Comment:
This is something Vector should do. Vector's goal is to be the only agent necessary to collect and ship all observability data.

How difficult would it be for K8s operators to skip deploying the `node_exporter` in favor of Vector?

Reply:
`node_exporter` has [more metrics(https://github.com/prometheus/node_exporter#enabled-by-default) than we currently support at [`host_metrics`](https://github.com/timberio/vector/blob/master/rfcs/2020-08-26-3191-host-metrics.md), so we might need to do some more work on the `host_metrics` before we'd be able to fully replace the `node_exporter`.
We can do this incrementally though, and start with shipping our `host_metrics` source as per #4163. For some users just gathering the `host_metrics` would be enough.